docs: update `no-restricted-imports` rule (#18015)
eslint/eslint: docs/src/rules/no-restricted-imports.md

## Options

This rule has both string and object options to specify the imported modules to restrict.

> Simplest way to specify a module that you want to restrict from being imported is to pass it's name as a string to the options:

Reviewer: ```suggestion
Using string options, you can specify the name of a module that you want to restrict from being imported as a value in the rule options array:
```

```json
"no-restricted-imports": ["error", "import1", "import2"]
```

Examples of **incorrect** code for string option:

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", "fs"]*/

import fs from 'fs';
```

:::

> Specified string can also restrict the module from being exported:

Reviewer: ```suggestion
String options also restrict the module from being exported, as in this example:
```

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", "fs"]*/

export { fs } from 'fs';
```

:::

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", "fs"]*/

export * from 'fs';
```

:::

Examples of **correct** code for string option:

::: correct { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", "fs"]*/

import crypto from 'crypto';
export { foo } from "bar";
```

:::

> You may also specify a custom message for a particular module simply using `name` and `message` properties inside an object, where the value of the `name` property will be the name of the module and `message` property will contain the custom message. The custom message will be appended to the default error message:

Reviewer: ```suggestion
You may also specify a custom message for a particular module using the `name` and `message` properties inside an object, where the value of the `name` is the name of the module and `message` property contains the custom message. (The custom message is appended to the default error message from the rule.)
```

```json
"no-restricted-imports": ["error", {
    "name": "import-foo",
    "message": "Please use import-bar instead."
}, {
    "name": "import-baz",
    "message": "Please use import-quux instead."
}]
```

Examples of **incorrect** code for string option:

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", {
    "name": "disallowed-import",
    "message": "Please use 'allowed-import' instead"
}]*/

import foo from 'disallowed-import';
```

:::

### paths

> This is an object option whose value is an array and the name of the module you want to restrict can be specified inside this array, same as in string option above.

Reviewer: ```suggestion
This is an object option whose value is an array containing the names of the modules you want to restrict.
```

```json
"no-restricted-imports": ["error", { "paths": ["import1", "import2"] }]
```

Examples of **incorrect** code for `paths`:

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { "paths": ["cluster"] }]*/

import cluster from 'cluster';
```

:::

> Custom messages for a particular module can also be specified in `paths` array option using objects as above.

Reviewer: ```suggestion
Custom messages for a particular module can also be specified in `paths` array using objects with `name` and `message`.
```

```json
"no-restricted-imports": ["error", {
    "paths": [{
        "name": "import-foo",
        "message": "Please use import-bar instead."
    }, {
        "name": "import-baz",
        "message": "Please use import-quux instead."
    }]
}]
```

#### importNames

> This option in `paths` is an array and can be used to specify certain imports to restrict that are being exported from a module. Import names specified inside `paths` array looks for the module specified in the `name` property of corresponding object, so it is required to specify the `name` key with both `importNames` and `message` options.

Reviewer: ```suggestion
This option in `paths` is an array and can be used to specify the names of certain bindings exported from a module. Import names specified inside `paths` array affect the module specified in the `name` property of corresponding object, so it is required to specify the `name` key with both `importNames` and `message` options.
```

Question: is `message` actually required when using `importNames`? It doesn't seem like it should be so just double-checking.

Author: > it is required to specify the `name` key with both `importNames` and `message` options.

For me this line is saying when using `importNames` or `message` property, then it is require to use `name` property first. I think i should clear it's language a bit.

> Passing `"default"` string inside the `importNames` array will restrict the import that is being exported by default by the module.

Reviewer: ```suggestion
Specifying `"default"` inside `importNames` array will restrict the default export from being imported.
```

:::

Examples of **correct** code for `importNames` in `paths`:

> Other strings doesn't report default exported import even if the names are same:

Reviewer: I'm not entirely sure what this means. Maybe this?
```suggestion
If the local name assigned to a default export is the same as a string in `importNames`, this will not cause an error.
```

Author: Exactly this.

::: correct { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { paths: [{ name: "foo", importNames: ["DisallowedObject"] }] }]*/

import DisallowedObject from "foo"
```

:::

::: correct { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { paths: [{
    name: "foo",
    importNames: ["DisallowedObject"],
    message: "Please import 'DisallowedObject' from '/bar/baz/' instead."
}]}]*/

import { AllowedObject as DisallowedObject } from "foo";
```

:::

### patterns

> This is also an object option whose value is an array. This option comes handy when you need to restrict the modules present inside a perticular directory. Simply, you can specify the modules inside the directories you want to restrict as string in `gitignore-style` patterns.

Reviewer: ```suggestion
This is also an object option whose value is an array. This option allows you to specify multiple modules to restrict using `gitignore`-style patterns.
```

> Since the patterns follows the `gitignore-style` then if you want to reinclude any particular module this can be done by prefixing a negation (`!`) mark in front of the pattern and it's good to keep the negation patterns in last as `gitignore-style` follows order:

Reviewer: ```suggestion
Because the patterns follow `gitignore`-style, if you want to reinclude any particular module this can be done by prefixing a negation (`!`) mark in front of the pattern. (Negated patterns should come last in the array because order is important.)
```

```json
"no-restricted-imports": ["error", {
    "patterns": ["import1/private/*", "import2/*", "!import2/good"]
}]
```

Examples of **incorrect** code for `pattern` option:

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { "patterns": ["lodash/*"] }]*/

import pick from 'lodash/pick';
```

:::

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { "patterns": ["lodash/*", "!lodash/pick"] }]*/

import pick from 'lodash/map';
```

:::

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { "patterns": ["import1/*", "!import1/private/*"] }]*/

import pick from 'import1/private/someModule';
```

:::

> In last example `"!import1/private/*"` is not reincluding the modules inside private directory because in `gitignore-style` negation mark (`!`) does not reinclude the files if it's parent directory is excluded by a pattern. In this case `import1/private` directory is already excluded by the `import1/*` pattern, but excluded directory can be reincluded like using `"!import1/private"`.

Reviewer: ```suggestion
In this example, `"!import1/private/*"` is not reincluding the modules inside `private` because the negation mark (`!`) does not reinclude the files if it's parent directory is excluded by a pattern. In this case, `import1/private` directory is already excluded by the `import1/*` pattern. (The excluded directory can be reincluded using `"!import1/private"`.)
```

Examples of **correct** code for `pattern` option:

::: correct { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { "patterns": ["crypto/*"] }]*/

import crypto from 'crypto';
```

:::

::: correct { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { "patterns": ["lodash/*", "!lodash/pick"] }]*/

import pick from 'lodash/pick';
```

:::

::: correct { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { "patterns": ["import1/*", "!import1/private"] }]*/

import pick from 'import1/private/someModule';
```

:::

#### group

> Alternatively you can use this object property which is an array to specify the `gitignore-style` patterns for restricting modules. Custom messages can also be specified like `paths`.
> 
> This is a required property if you are using object options inside the `patterns` array:

Reviewer: ```suggestion
The `patterns` array can also include objects. The `group` property is used to specify the `gitignore`-style patterns for restricting modules and the `message` property is used to specify a custom message.

The `group` property is required property when using objects inside the `patterns` array.
```

:::

#### caseSensitive

> This is a boolean option and allows the patterns specified in the `group` array to be case-sensitive when sets to `true`. Default is `false`:

Reviewer: ```suggestion
This is a boolean option and sets the patterns specified in the `group` array to be case-sensitive when `true`. Default is `false`.
```

:::

#### importNames

> `patterns` too has an `importNames` property and this works similar as in `paths`, it just looks for patterns specified inside the `group` array:

Reviewer: ```suggestion
You can also specify `importNames` on objects inside of `patterns`. In this case, the specified names are applied only to the specified `group`.
```

Author: Done! till here

:::

To restrict the use of all Node.js core imports (via <https://github.com/nodejs/node/tree/master/lib>):

```json
    "no-restricted-imports": ["error",
         "assert","buffer","child_process","cluster","crypto","dgram","dns","domain","events","freelist","fs","http","https","module","net","os","path","punycode","querystring","readline","repl","smalloc","stream","string_decoder","sys","timers","tls","tracing","tty","url","util","vm","zlib"
    ],
```

> 

Reviewer: I would leave this out. It's a very specific use case that it easy to figure out if you need it. (It would probably also be a better idea to use the `globals` package for this purpose.)

Author: i this means we can remove this section?

Reviewer: Yes, I think we can remove this section. It's not necessary.

Author: Done!
